File: app/src/main/java/io/github/samueljarosinski/ambilight/hue/HueController.java
package io.github.samueljarosinski.ambilight.hue;

import android.os.Build;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.philips.lighting.hue.sdk.wrapper.domain.Bridge;
import java.util.concurrent.TimeUnit;
import io.github.samueljarosinski.ambilight.hue.bridge.BridgeController;
import io.github.samueljarosinski.ambilight.hue.bridge.BridgeController.OnBridgeConnectedListener;
import io.github.samueljarosinski.ambilight.hue.lights.LightsController;
import timber.log.Timber;

public class HueController {

    public static final String APP_NAME = "Ambilight";
    public static final String DEVICE_NAME = Build.MODEL;
    public static final int HEARTBEAT_INTERVAL = (int) TimeUnit.SECONDS.toMillis(10);
    public static final int MIN_UPDATE_DELAY = 100;

    @NonNull
    private final BridgeController bridgeController;

    @Nullable
    private LightsController lightsController;

    @Nullable
    private OnHueConnectionListener onHueConnectionListener;

    public HueController () {
        bridgeController = new BridgeController();
    }

    public void start (@Nullable OnHueConnectionListener onHueConnectionListener) {
        Timber.d("Starting Hue.");

        this.onHueConnectionListener = onHueConnectionListener;

        bridgeController.connect(new OnBridgeConnectedListener() {

            @Override
            public void onBridgeConnected (@NonNull Bridge bridge) {
                lightsController = new LightsController(bridge);

                if (HueController.this.onHueConnectionListener != null) {
                    HueController.this.onHueConnectionListener.onConnected();
                }
            }

            @Override
            public void onConnectionError () {
                if (HueController.this.onHueConnectionListener != null) {
                    HueController.this.onHueConnectionListener.onConnectionError();
                }
            }
        });
    }

    public void stop () {
        Timber.d("Stopping Hue.");

        onHueConnectionListener = null;
        lightsController = null;
        bridgeController.disconnect();
    }

    public void setColor (@ColorInt int color) {
        if (lightsController != null) {
            lightsController.setColor(color);
        }
    }

    public interface OnHueConnectionListener {
        void onConnected ();
        void onConnectionError ();
    }

}

File: app/src/main/java/io/github/samueljarosinski/ambilight/hue/HueController.kt
package io.github.samueljarosinski.ambilight.hue

import android.os.Build
import androidx.annotation.ColorInt
import com.philips.lighting.hue.sdk.wrapper.domain.Bridge
import timber.log.Timber

val DEVICE_NAME: String = Build.MODEL
internal const val APP_NAME = "Ambilight"
internal const val HEARTBEAT_INTERVAL = 10000
const val MIN_UPDATE_DELAY = 100

class HueController {

    private val bridgeController: BridgeController = BridgeController()
    private var lightsController: LightsController? = null
    private var hueConnectionListener: OnHueConnectionListener? = null

    fun start(onHueConnectionListener: OnHueConnectionListener?) {
        Timber.d("Starting Hue.")

        hueConnectionListener = onHueConnectionListener
        bridgeController.connect(object : OnBridgeConnectedListener {

            override fun onNoBridgesFound() {
                hueConnectionListener?.onNoBridgesFound()
            }

            override fun onNotAuthenticated() {
                hueConnectionListener?.onNotAuthenticated()
            }

            override fun onBridgeConnected(bridge: Bridge) {
                lightsController = LightsController(bridge)
                hueConnectionListener?.onConnected()
            }

            override fun onConnectionError() {
                hueConnectionListener?.onConnectionError()
            }
        })
    }

    fun stop() {
        Timber.d("Stopping Hue.")

        hueConnectionListener = null
        lightsController = null
        bridgeController.disconnect()
    }

    fun setColor(@ColorInt color: Int) {
        lightsController?.setColor(color)
    }
}

interface OnHueConnectionListener {
    fun onNoBridgesFound()
    fun onNotAuthenticated()
    fun onConnected()
    fun onConnectionError()
}
